### PR TITLE
Create a dockerfile for repeatable releases

### LINKED ISSUE #15: Make a basic publish-new-release.dockerfile
Building a release-generating-dockerfile is probably the easiest short-term way to get repeatable "semi-local" builds, deployable by more than just one person. Long term of course, we'd favor git actions or equivalent, but this will only make that step easier.

This is a good step toward #5

### PR DESCRIPTION
**Pull request by snickell**
Fixes #15 

**UPDATE:** this is now working, and has released rc.1 and rc.2, which both appear to be valid

It can be used as follows, assuming your npm account has permissions to publish to maplibre-gl:
```
./build/publish-release.sh
```

Then it will:
1. Prompt you for the new version number
2. Build a number of asset types, and test maplibre-gl
3. Diff the tarball with the previous release of `maplibre-gl` on npm and ask you to confirm (makes it easy to spot build errors, etc), note you'll need to type y or n and press enter at this phase 🤷‍♂️
4. Publish the new version to npm
5. Update your package.json and git tags on the host machine even though the thing executes in docker 🧙🏿‍♂️
6. Give you a link to the new release

You can also run alternate yarn targets as follows, e.g. to run tests:
```
./build/publish-release.sh "yarn test"
```

Slight variations on this in the future should make CI integrations smooth and easy.

### CODE REVIEW

**Comment by snickell** [comment]
This is currently very WIP and doesn't succeed yet, but good to get where others can see it

File: build/publish-release.dockerfile [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ADD . maplibre-gl-js
+# OR Checkout straight from the source
+# RUN git clone https://github.com/maplibre/maplibre-gl-js.git --single-branch maplibre-gl-js
+
+WORKDIR /maplibre-gl-js
+
+RUN \
+  yarn install && \
+  yarn prepublishOnly && \
+  npm publish
```
Comment (nyurik) [inline build/publish-release.dockerfile:35]: not even sure how `npm login` would work from inside docker tbh. Also, would be good if the `build` docker could be shared with the `publish` docker, as most of their content would be identical except for the last few lines. Perhaps we shouldn't try to do this as dockerfile, but instead have a few scripts in the `scripts/` dir, and have just one generic docker that essentially just sets up node+yarn+other required apt installs, but does not copy code or compile it. Instead, the scripts would run inside the docker to do the needed tasks, and will take params as needed.  Also, another script would be the "docker run" script, which will pass in npm login information from the host, plus will take the name of the script to run inside docker.

Comment (snickell) [reply build/publish-release.dockerfile]: So far the primary issue has been version compat with npms and various system packages, so as things are a little ad-hoc intiially, I think Docker is the easiest approach to that I know of to make sure that nobody who pushes a release, it'll be the same as somebody else doing it.

Comment (snickell) [reply build/publish-release.dockerfile]: See update to the PR

**Comment by snickell** [comment]
Getting closer, this runs the build phase and many/most of the tests, but fails out right before publishing:
```
➜  build git:(seth/release-dockerfile) ✗ ./publish-release.sh
+ docker build -t maplibre-gl/publish-release -f ./publish-release.dockerfile .
[+] Building 41.8s (9/9) FINISHED                                                
 => [internal] load build definition from publish-release.dockerfile        0.0s
 => => transferring dockerfile: 1.05kB                                      0.0s
 => [internal] load .dockerignore                                           0.0s
 => => transferring context: 2B                                             0.0s
 => [internal] load metadata for docker.io/library/node:15                  0.0s
 => [1/4] FROM docker.io/library/node:15                                    0.0s
 => [internal] load build context                                           0.0s
 => => transferring context: 44B                                            0.0s
 => CACHED [2/4] RUN   apt update -y &&   apt install -y apt-transport-htt  0.0s
 => [3/4] RUN   apt install -y     git     node-gyp xserver-xorg-dev libx  36.4s
 => [4/4] ADD publish-release-cmd.sh .                                      0.9s
 => exporting to image                                                      4.5s 
 => => exporting layers                                                     4.4s 
 => => writing image sha256:af0cd91c86ead16eaa3c716987cceec6b95da2f2fb1448  0.0s 
 => => naming to docker.io/maplibre-gl/publish-release                      0.0s 
+ docker run -it --env NPM_TOKEN maplibre-gl/publish-release                     
+ git clone https://github.com/maplibre/maplibre-gl-js.git --single-branch
Cloning into 'maplibre-gl-js'...
remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 83074 (delta 4), reused 2 (delta 0), pack-reused 83059
Receiving objects: 100% (83074/83074), 153.68 MiB | 15.91 MiB/s, done.
Resolving deltas: 100% (55771/55771), done.
+ cd maplibre-gl-js
+ yarn install
yarn install v1.22.5
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@2.1.2: The platform "linux" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.12: The platform "linux" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning " > @mapbox/mapbox-gl-supported@1.5.0" has unmet peer dependency "mapbox-gl@>=0.32.1 <2.0.0".
warning " > @mapbox/mapbox-gl-rtl-text@0.2.3" has unmet peer dependency "mapbox-gl@>=0.32.1 <2.0.0".
[5/5] Building fresh packages...
Done in 415.89s.
+ yarn prepublishOnly
yarn run v1.22.5
$ run-s prepare-publish build-flow-types build-dev build-prod-min build-prod build-csp build-css build-style-spec test-build diff-tarball
$ git clean -fdx && yarn install
Removing node_modules/
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@2.1.2: The platform "linux" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.12: The platform "linux" is incompatible with this module.
info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning " > @mapbox/mapbox-gl-supported@1.5.0" has unmet peer dependency "mapbox-gl@>=0.32.1 <2.0.0".
warning " > @mapbox/mapbox-gl-rtl-text@0.2.3" has unmet peer dependency "mapbox-gl@>=0.32.1 <2.0.0".
[5/5] Building fresh packages...
$ mkdir -p dist && cp build/mapbox-gl.js.flow dist/mapbox-gl.js.flow && cp build/mapbox-gl.js.flow dist/mapbox-gl-dev.js.flow
$ rollup -c --environment BUILD:dev

src/index.js, src/source/worker.js → rollup/build/mapboxgl...
(!) Circular dependencies
src/util/ajax.js -> src/util/mapbox.js -> src/util/ajax.js
src/style-spec/expression/parsing_context.js -> src/style-spec/expression/compound_expression.js -> src/style-spec/expression/parsing_context.js
src/style-spec/validate/validate.js -> src/style-spec/validate/validate_function.js -> src/style-spec/validate/validate.js
...and 10 more
created rollup/build/mapboxgl in 12.2s

rollup/mapboxgl.js → dist/mapbox-gl-dev.js...
created dist/mapbox-gl-dev.js in 3.3s
$ rollup -c --environment BUILD:production,MINIFY:true

src/index.js, src/source/worker.js → rollup/build/mapboxgl...
(!) Circular dependencies
src/util/ajax.js -> src/util/mapbox.js -> src/util/ajax.js
src/style-spec/expression/parsing_context.js -> src/style-spec/expression/compound_expression.js -> src/style-spec/expression/parsing_context.js
src/style-spec/validate/validate.js -> src/style-spec/validate/validate_function.js -> src/style-spec/validate/validate.js
...and 10 more
created rollup/build/mapboxgl in 34.8s

rollup/mapboxgl.js → dist/mapbox-gl.js...
created dist/mapbox-gl.js in 3.1s
$ rollup -c --environment BUILD:production

src/index.js, src/source/worker.js → rollup/build/mapboxgl...
(!) Circular dependencies
src/util/ajax.js -> src/util/mapbox.js -> src/util/ajax.js
src/style-spec/expression/parsing_context.js -> src/style-spec/expression/compound_expression.js -> src/style-spec/expression/parsing_context.js
src/style-spec/validate/validate.js -> src/style-spec/validate/validate_function.js -> src/style-spec/validate/validate.js
...and 10 more
created rollup/build/mapboxgl in 21.7s

rollup/mapboxgl.js → dist/mapbox-gl-unminified.js...
created dist/mapbox-gl-unminified.js in 3.3s
$ rollup -c rollup.config.csp.js

src/index.js → dist/mapbox-gl-csp.js...
(!) Circular dependencies
src/util/ajax.js -> src/util/mapbox.js -> src/util/ajax.js
src/style-spec/expression/parsing_context.js -> src/style-spec/expression/compound_expression.js -> src/style-spec/expression/parsing_context.js
src/style-spec/validate/validate.js -> src/style-spec/validate/validate_function.js -> src/style-spec/validate/validate.js
...and 10 more
created dist/mapbox-gl-csp.js in 40.2s

src/source/worker.js → dist/mapbox-gl-csp-worker.js...
(!) Circular dependencies
src/style-spec/expression/parsing_context.js -> src/style-spec/expression/compound_expression.js -> src/style-spec/expression/parsing_context.js
src/style-spec/validate/validate.js -> src/style-spec/validate/validate_function.js -> src/style-spec/validate/validate.js
src/style-spec/validate/validate.js -> src/style-spec/validate/validate_function.js -> src/style-spec/validate/validate_object.js -> src/style-spec/validate/validate.js
...and 10 more
created dist/mapbox-gl-csp-worker.js in 21.6s
$ postcss -o dist/mapbox-gl.css src/css/mapbox-gl.css
Browserslist: caniuse-lite is outdated. Please run the following command: `npx browserslist --update-db`
$ cd src/style-spec && npm run build && cd ../.. && mkdir -p dist/style-spec && cp src/style-spec/dist/* dist/style-spec

> @mapbox/mapbox-gl-style-spec@13.18.0-dev build
> ../../node_modules/.bin/rollup -c && ../../node_modules/.bin/rollup -c --environment esm


/maplibre-gl-js/src/style-spec/style-spec.js → dist/index.js...
(!) Circular dependencies
expression/parsing_context.js -> expression/compound_expression.js -> expression/parsing_context.js
validate/validate.js -> validate/validate_function.js -> validate/validate.js
validate/validate.js -> validate/validate_function.js -> validate/validate_object.js -> validate/validate.js
...and 6 more
created dist/index.js in 6.2s

/maplibre-gl-js/src/style-spec/style-spec.js → dist/index.es.js...
(!) Circular dependencies
expression/parsing_context.js -> expression/compound_expression.js -> expression/parsing_context.js
validate/validate.js -> validate/validate_function.js -> validate/validate.js
validate/validate.js -> validate/validate_function.js -> validate/validate_object.js -> validate/validate.js
...and 6 more
created dist/index.es.js in 6.1s
$ build/run-tap --no-coverage test/build/**/*.test.js
/maplibre-gl-js/node_modules/@mapbox/flow-remove-types/index.js:1
SyntaxError: Error parsing /maplibre-gl-js/node_modules/babylon/package.json: Unexpected token , in JSON at position 726
    at JSON.parse (<anonymous>) {
  path: '/maplibre-gl-js/node_modules/babylon/package.json'
}
/maplibre-gl-js/node_modules/@mapbox/flow-remove-types/index.js:1
SyntaxError: Error parsing /maplibre-gl-js/node_modules/babylon/package.json: Unexpected token , in JSON at position 726
    at JSON.parse (<anonymous>) {
  path: '/maplibre-gl-js/node_modules/babylon/package.json'
}
test/build/dev.test.js ................................ 0/1
  not ok test/build/dev.test.js
    timeout: 30000
    file: test/build/dev.test.js
    childId: 0
    command: /usr/local/bin/node
    args:
      - '-r'
      - '@mapbox/flow-remove-types/register'
      - '-r'
      - esm
      - test/build/dev.test.js
    stdio:
      - 0
      - pipe
      - 2
    cwd: /maplibre-gl-js
    exitCode: 1

/maplibre-gl-js/node_modules/@mapbox/flow-remove-types/index.js:1
SyntaxError: Error parsing /maplibre-gl-js/node_modules/babylon/package.json: Unexpected token , in JSON at position 726
    at JSON.parse (<anonymous>) {
  path: '/maplibre-gl-js/node_modules/babylon/package.json'
}
test/build/min.test.js ................................ 0/1
  not ok test/build/min.test.js
    timeout: 30000
    file: test/build/min.test.js
    childId: 1
    command: /usr/local/bin/node
    args:
      - '-r'
      - '@mapbox/flow-remove-types/register'
      - '-r'
      - esm
      - test/build/min.test.js
    stdio:
      - 0
      - pipe
      - 2
    cwd: /maplibre-gl-js
    exitCode: 1

test/build/style-spec.test.js ......................... 0/1
  not ok test/build/style-spec.test.js
    timeout: 30000
    file: test/build/style-spec.test.js
    childId: 2
    command: /usr/local/bin/node
    args:
      - '-r'
      - '@mapbox/flow-remove-types/register'
      - '-r'
      - esm
      - test/build/style-spec.test.js
    stdio:
      - 0
      - pipe
      - 2
    cwd: /maplibre-gl-js
    exitCode: 1

total ................................................. 0/3
  

  0 passing (896.451ms)
  3 failing

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "test-build" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
➜  build git:(seth/release-dockerfile) 
```

**Comment by snickell** [comment]
Well I'm a little stumped on this one, might be worth fresh eyes taking a look (@nyurik ?)

Context:
```
/src/node_modules/@mapbox/flow-remove-types/index.js:1
SyntaxError: Error parsing /src/node_modules/babylon/package.json: Unexpected token , in JSON at position 726
    at JSON.parse (<anonymous>) {
  path: '/src/node_modules/babylon/package.json'
}
test/build/dev.test.js ................................ 0/1
  not ok test/build/dev.test.js
    timeout: 30000
    file: test/build/dev.test.js
    childId: 0
    command: /usr/local/bin/node
    args:
      - '-r'
      - '@mapbox/flow-remove-types/register'
      - '-r'
      - esm
      - test/build/dev.test.js
```

1. All the tests are run by mapbox-gl-js are run using `./build/run-node` (see https://github.com/maplibre/maplibre-gl-js/blob/seth/release-dockerfile/build/run-node#L4:L4), which is just:
```
node -r @mapbox/flow-remove-types/register -r esm ${@}
```
2. `run-node` is necessary to strip the flow types out of files starting with a `//@flow` comment, its basically a type preprocessor system (🙀)
3. In this case, `@mapbox/flow-remove-types` appears (and sadly, doesn't print ANY context or useful backtrace here how it followed this path) to be following the deps chain `test/build/*.test.js` until it tries to parse `node_modules/babylon/package.json`, which is probably import/required somewhere by somebody.
4. `@mapbox/flow-remove-types` probably (right?) isn't supposed to try to parse JSON, but it doesn't know that, so it dies (`position 726` in babylon/package.json is End Of File)
5. `@mapbox/flow-remove-types` is a very old fork of flow-remove-types, they forked to release some stuff on head, but I suspect flow-remove-types didn't ever merge their changes. I did try using a current flow-remove-types, but it broke much earlier, and I don't think mapbox flow is compatible.

So where do we break the chain to keep `@mapbox/flow-remove-types` stripping flow types (necessary to run tests), but not following a crazy dep chain until it tries to parse JSON in the form of `babylon/package.json`?

A good first step would be figuring out the dep chain that's being followed to land in babylon/package.json, starting from https://github.com/maplibre/maplibre-gl-js/blob/seth/release-dockerfile/test/build/dev.test.js#L1:L2

**Comment by snickell** [comment]
If anyone wants to carry this further and reproduce, it should be very simple:

1. Check out the `seth/release-dockerfile` branch on this repo
2. `bash ./build/publish-release.sh`

In about 5 minutes, you should see it fail on this error.

**Comment by snickell** [comment]
One thing is that this is just `test-build`, which is basically a few simple structure tests for builds (out of the many many other tests which do work). It may be worth blocking on, but I suspect the right thing to do here is to disable this package.json script, get repeatable releases going, and file an issue to get this structure test working again (it looks like it could be anywhere from really tricky to totally trivial).

**Comment by JamesLMilner** [comment]
@snickell thinking about it I don't think `remove-flow-types` should be trying to strip anything in node_modules at all as that will be built files for importing? I would have to look into it but maybe there's an easy way to pass excluded files/directories to `remove-flow-types`. Also it would probably be worth looking at converting to TypeScript further down the line as it seems like the vast majority of projects have moved to TypeScript now (larger community/support/backing etc)

**Comment by snickell** [comment]
@JamesLMilner yeah, I was hoping we could exclude and not even try to track down why its starting to look there

**Comment by snickell** [comment]
**Successful deploy to NPM of https://www.npmjs.com/package/maplibre-gl/v/1.13.0-rc.1 from the dockerfile!**

 After disabling `test-build`, we got a successful build!

**Comment by snickell** [comment]
Aha! Useful tip from @jdesboeufs:
```
Jérôme Desboeufs  6 minutes ago
This issue only occur on Node.js 12+
Jérôme Desboeufs  6 minutes ago
If you run tests on Node.js 10 it works…
Jérôme Desboeufs  4 minutes ago
Most of Mapbox repositories only works on Node10.
```

**Comment by snickell** [comment]
OK! It now supports (optional) variable yarn commands to start enabling other types of CI integration, e.g.:
```
./build/publish-release.sh "yarn test"
```

**Comment by snickell** [comment]
OK, I've updated the docs for making a release https://github.com/maplibre/maplibre-gl-js/blob/seth/release-dockerfile/docs/making-a-release.md#L8:L8

It would be great if somebody else with npmjs permission would try them out and try making a release? This will give us confidence and remove me as a potential bottleneck. @nyurik , wanna try to release rc.4?

**Pull request merged**